File: languages/c/k-nucleotide.c
// k-nucleotide: count the frequency of every length-K substring (k-mer) of a
// deterministically generated DNA sequence, using a hash map, then reduce the map to
// one order-independent checksum.
//
// This is the associative-container axis of the suite. C has no standard hash map, so
// we hand-roll an open-addressing table with a general string hash (FNV-1a) and linear
// probing - a real hash map, NOT direct-addressing on the small key space (which would
// be an unfair shortcut). Every other language uses its built-in map.
//
// Everything is integer-deterministic (no floating point): the sequence comes from an
// integer LCG, and the checksum is sum over map entries of encode(kmer)*count mod P,
// which is independent of the map's hash function and iteration order.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define K 8
#define P 1000000007L
#define IM 139968
#define IA 3877
#define IC 29573
#define HBITS 18
#define HSIZE (1 << HBITS)
#define HMASK (HSIZE - 1)

static char *gen(int L) {
    char *s = malloc((size_t)L + 1);
    long seed = 42;
    for (int i = 0; i < L; i++) {
        seed = (seed * IA + IC) % IM;
        s[i] = seed < 42000 ? 'A' : seed < 70000 ? 'C' : seed < 98000 ? 'G' : 'T';
    }
    s[L] = '\0';
    return s;
}

static char htkey[HSIZE][K];
static long htcnt[HSIZE];
static int  htused[HSIZE];

static unsigned long fnv(const char *p) {
    unsigned long h = 1469598103934665603UL;
    for (int i = 0; i < K; i++) { h ^= (unsigned char)p[i]; h *= 1099511628211UL; }
    return h;
}

static void add(const char *kmer) {
    unsigned long h = fnv(kmer) & HMASK;
    while (htused[h]) {
        if (memcmp(htkey[h], kmer, K) == 0) { htcnt[h]++; return; }
        h = (h + 1) & HMASK;
    }
    htused[h] = 1;
    memcpy(htkey[h], kmer, K);
    htcnt[h] = 1;
}

int main(int argc, char **argv) {
    int L = argc > 1 ? atoi(argv[1]) : 100000;
    char *s = gen(L);
    for (int i = 0; i + K <= L; i++) add(s + i);
    long acc = 0;
    for (int b = 0; b < HSIZE; b++) {
        if (!htused[b]) continue;
        long e = 0;
        for (int j = 0; j < K; j++) {
            char c = htkey[b][j];
            int code = c == 'A' ? 0 : c == 'C' ? 1 : c == 'G' ? 2 : 3;
            e = e * 4 + code;
        }
        acc = (acc + e * htcnt[b]) % P;
    }
    printf("%ld\n", acc);
    printf("k-nucleotide(%d)\n", L);
    free(s);
    return 0;
}

File: languages/ruby/k-nucleotide.rb
# k-nucleotide: count the frequency of every length-K substring (k-mer) of a
# deterministically generated DNA sequence using the built-in Hash, then reduce the
# map to one order-independent checksum.
#
# Everything is integer-deterministic (no floating point): the sequence comes from an
# integer LCG, and the checksum is sum over map entries of encode(kmer)*count mod P,
# which is independent of the Hash's iteration order. The key is the k-mer string -
# no direct-addressing shortcut on the small 4^8 key space.

K = 8
P = 1000000007
IM = 139968
IA = 3877
IC = 29573

CODE = { "A" => 0, "C" => 1, "G" => 2, "T" => 3 }

def gen(length)
  seed = 42
  chars = []
  length.times do
    seed = (seed * IA + IC) % IM
    chars <<
      if seed < 42000
        "A"
      elsif seed < 70000
        "C"
      elsif seed < 98000
        "G"
      else
        "T"
      end
  end
  chars.join
end

def k_nucleotide(length)
  s = gen(length)

  counts = Hash.new(0)
  i = 0
  last = length - K
  while i <= last
    counts[s[i, K]] += 1
    i += 1
  end

  acc = 0
  counts.each do |kmer, count|
    e = 0
    kmer.each_char { |ch| e = e * 4 + CODE[ch] }
    acc = (acc + e * count) % P
  end
  acc
end

if __FILE__ == $PROGRAM_NAME
  n = ARGV[0] ? ARGV[0].to_i : 200000
  puts k_nucleotide(n)
  puts "k-nucleotide(#{n})"
end
